### ISSUE TITLE
Loki: Labels with specifying only key causes invalid memory access

### ISSUE THREAD

**Issue by toshipp** [fluent/fluent-bit#3279]
## Bug Report

**Describe the bug**
Using loki output plugin with key only label causes invalid memory access.
Sometimes, fluentbit is crashed with sigsegv.

**To Reproduce**
Use following configuration and run fluentbit with valgrind.

```
    [INPUT]
        Name tail
        Path /var/log/containers/*.log
        Parser cri
        Tag kube.*
        Buffer_Max_Size 1MB
        Mem_Buf_Limit 128MB
        Skip_Long_Lines On
        DB /run/fluent-bit/tail.db
    [FILTER]
        Name kubernetes
        Match kube.*
        # Merge_Log On
        # Keep_Log Off
        K8S-Logging.Parser On
        K8S-Logging.Exclude On
        Annotations Off
    [OUTPUT]
        Name loki
        Match kube.*
        Host logging-loki
        labels $pod_name
```

Valgrind reports invalid memory reads and writes.

```
==1== Invalid read of size 1
==1==    at 0x20F2EC: pack_label_key (loki.c:281)
==1==    by 0x20F522: pack_labels (loki.c:324)
==1==    by 0x211435: loki_compose_payload (loki.c:1024)
==1==    by 0x2115A5: cb_loki_flush (loki.c:1062)
==1==    by 0x16F183: output_pre_cb_flush (flb_output.h:466)
==1==    by 0x48794A: co_init (amd64.c:117)
==1==    by 0x72656B3A725F6D64: ???
==1==    by 0x30733A745F6C656D: ???
==1==  Address 0xd42032e is 8,190 bytes inside a block of size 8,192 free'd
==1==    at 0x483DFAF: realloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==1==    by 0x20E655: msgpack_sbuffer_write (sbuffer.h:81)
==1==    by 0x20E473: msgpack_pack_str_body (pack_template.h:784)
==1==    by 0x20F2CD: pack_label_key (loki.c:278)
==1==    by 0x20F522: pack_labels (loki.c:324)
==1==    by 0x211435: loki_compose_payload (loki.c:1024)
==1==    by 0x2115A5: cb_loki_flush (loki.c:1062)
==1==    by 0x16F183: output_pre_cb_flush (flb_output.h:466)
==1==    by 0x48794A: co_init (amd64.c:117)
==1==    by 0x72656B3A725F6D64: ???
==1==    by 0x30733A745F6C656D: ???
==1==  Block was alloc'd at
==1==    at 0x483B723: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==1==    by 0x483E017: realloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==1==    by 0x20E655: msgpack_sbuffer_write (sbuffer.h:81)
==1==    by 0x20E275: msgpack_pack_map (pack_template.h:745)
==1==    by 0x21120F: loki_compose_payload (loki.c:962)
==1==    by 0x2115A5: cb_loki_flush (loki.c:1062)
==1==    by 0x16F183: output_pre_cb_flush (flb_output.h:466)
==1==    by 0x48794A: co_init (amd64.c:117)
==1==    by 0x72656B3A725F6D64: ???
==1==    by 0x30733A745F6C656D: ???
==1==
==1== Invalid read of size 1
==1==    at 0x20F313: pack_label_key (loki.c:281)
==1==    by 0x20F522: pack_labels (loki.c:324)
==1==    by 0x211435: loki_compose_payload (loki.c:1024)
==1==    by 0x2115A5: cb_loki_flush (loki.c:1062)
==1==    by 0x16F183: output_pre_cb_flush (flb_output.h:466)
==1==    by 0x48794A: co_init (amd64.c:117)
==1==    by 0x72656B3A725F6D64: ???
==1==    by 0x30733A745F6C656D: ???
==1==  Address 0xd420330 is 0 bytes after a block of size 8,192 free'd
==1==    at 0x483DFAF: realloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==1==    by 0x20E655: msgpack_sbuffer_write (sbuffer.h:81)
==1==    by 0x20E473: msgpack_pack_str_body (pack_template.h:784)
==1==    by 0x20F2CD: pack_label_key (loki.c:278)
==1==    by 0x20F522: pack_labels (loki.c:324)
==1==    by 0x211435: loki_compose_payload (loki.c:1024)
==1==    by 0x2115A5: cb_loki_flush (loki.c:1062)
==1==    by 0x16F183: output_pre_cb_flush (flb_output.h:466)
==1==    by 0x48794A: co_init (amd64.c:117)
==1==    by 0x72656B3A725F6D64: ???
==1==    by 0x30733A745F6C656D: ???
==1==  Block was alloc'd at
==1==    at 0x483B723: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==1==    by 0x483E017: realloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==1==    by 0x20E655: msgpack_sbuffer_write (sbuffer.h:81)
==1==    by 0x20E275: msgpack_pack_map (pack_template.h:745)
==1==    by 0x21120F: loki_compose_payload (loki.c:962)
==1==    by 0x2115A5: cb_loki_flush (loki.c:1062)
==1==    by 0x16F183: output_pre_cb_flush (flb_output.h:466)
==1==    by 0x48794A: co_init (amd64.c:117)
==1==    by 0x72656B3A725F6D64: ???
==1==    by 0x30733A745F6C656D: ???
==1==
==1== Invalid write of size 1
==1==    at 0x20F327: pack_label_key (loki.c:282)
==1==    by 0x20F522: pack_labels (loki.c:324)
==1==    by 0x211435: loki_compose_payload (loki.c:1024)
==1==    by 0x2115A5: cb_loki_flush (loki.c:1062)
==1==    by 0x16F183: output_pre_cb_flush (flb_output.h:466)
==1==    by 0x48794A: co_init (amd64.c:117)
==1==    by 0x72656B3A725F6D64: ???
==1==    by 0x30733A745F6C656D: ???
==1==  Address 0xd420332 is 2 bytes after a block of size 8,192 free'd
==1==    at 0x483DFAF: realloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==1==    by 0x20E655: msgpack_sbuffer_write (sbuffer.h:81)
==1==    by 0x20E473: msgpack_pack_str_body (pack_template.h:784)
==1==    by 0x20F2CD: pack_label_key (loki.c:278)
==1==    by 0x20F522: pack_labels (loki.c:324)
==1==    by 0x211435: loki_compose_payload (loki.c:1024)
==1==    by 0x2115A5: cb_loki_flush (loki.c:1062)
==1==    by 0x16F183: output_pre_cb_flush (flb_output.h:466)
==1==    by 0x48794A: co_init (amd64.c:117)
==1==    by 0x72656B3A725F6D64: ???
==1==    by 0x30733A745F6C656D: ???
==1==  Block was alloc'd at
==1==    at 0x483B723: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==1==    by 0x483E017: realloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==1==    by 0x20E655: msgpack_sbuffer_write (sbuffer.h:81)
==1==    by 0x20E275: msgpack_pack_map (pack_template.h:745)
==1==    by 0x21120F: loki_compose_payload (loki.c:962)
==1==    by 0x2115A5: cb_loki_flush (loki.c:1062)
==1==    by 0x16F183: output_pre_cb_flush (flb_output.h:466)
==1==    by 0x48794A: co_init (amd64.c:117)
==1==    by 0x72656B3A725F6D64: ???
==1==    by 0x30733A745F6C656D: ???
==1==
```

And Sometimes, fluentbit is died by sigsegv.

**Expected behavior**

No invalid memory access, No sigsegv.

**Your Environment**

* Version used: 1.7.2
* Configuration:
* Environment name and version (e.g. Kubernetes? What version?): kubernetes 1.19
* Server type and version: Qemu VM
* Operating System and version: 5.10.19-flatcar
* Filters and plugins: tail input, kubernetes filter, loki output

**Additional context**

My investigation is here.
The first time, `p` is made from `mp_sbuf`.
https://github.com/fluent/fluent-bit/blob/v1.7.2/plugins/out_loki/loki.c#L275
And then, `msgpack_pack_str_body` is called. It uses `realloc` and internal buffer shared with `mp_sbuf`
is changed. so `p` would be invalid.
Finally, memory contents referenced by `p` may be changed. The memory is already freed, so it is invalid memory access.
https://github.com/fluent/fluent-bit/blob/v1.7.2/plugins/out_loki/loki.c#L280-L284

`pack_label_key` is used for key only label.
If I use `Labels pod=$pod_name`, the problem never happens.

**Comment by JeffLuoo** [comment]
Seeing similar errors:
```
[2021/04/12 20:51:58] [engine] caught signal (SIGSEGV)
#0  0x7f268449a820      in  ???() at ???:0
#1  0x55c28d6220ed      in  msgpack_sbuffer_write() at lib/msgpack-c/include/msgpack/sbuffer.h:88
#2  0x55c28d621f52      in  msgpack_pack_str_body() at lib/msgpack-c/include/msgpack/pack_template.h:784
#3  0x55c28d6226d3      in  flb_metrics_dump_values() at src/flb_metrics.c:224
#4  0x55c28d6230d7      in  collect_outputs() at src/flb_metrics_exporter.c:139
#5  0x55c28d62321c      in  collect_metrics() at src/flb_metrics_exporter.c:167
#6  0x55c28d62337d      in  flb_me_fd_event() at src/flb_metrics_exporter.c:220
#7  0x55c28d60cdf0      in  flb_engine_handle_event() at src/flb_engine.c:359
#8  0x55c28d60cdf0      in  flb_engine_start() at src/flb_engine.c:613
#9  0x55c28d5f2cd2      in  flb_lib_worker() at src/flb_lib.c:493
#10 0x7f2684d56fa2      in  ???() at ???:0
#11 0x7f26844374ce      in  ???() at ???:0
#12 0xffffffffffffffff  in  ???() at ???:0
```

**Comment by toshipp** [comment]
still remains, IMO.

**Comment by toshipp** [comment]
not fixed yet

**Comment by nokute78** [comment]
I merged #3796 which is similar to #3847.

**Comment by toshipp** [comment]
Thanks to #3839, this is resolved.
